From 8df4ad6c7e42a1d3e0753716d62efa35b7eb6891 Mon Sep 17 00:00:00 2001
From: "M. Vefa Bicakci" <vefa.bicakci@windriver.com>
Date: Fri, 4 Feb 2022 18:27:53 -0500
Subject: [PATCH] bnxt_re/Makefile: Adapt to mlnx-ofa_kernel for StarlingX

This patch adapts the bnxt_re kernel module's Makefile to
mlnx-ofa_kernel for Debian-based StarlingX.

Here is a summary of the changes:

- Unlike CentOS-based StarlingX, the KDIR variable is no longer needed
  to be specified explicitly to make, because the symbolic links
  /lib/modules/<kver>/{build,source} are sufficient to detect the
  correct directories on Debian. However, this leaves the KDIR variable
  undefined, which is problematic as this variable is used later in the
  Makefile. Fix this by pointing KDIR to the "build" symbolic link
  (i.e., "$(LINUX)").

- To target Mellanox's OFED distribution instead of compat-rdma, the
  references to "/usr/src/compat-rdma" are replaced with
  "/usr/src/ofa_kernel".

- Mellanox's OFED distribution has version numbers in the form 5.x, so
  the Makefile version match patterns are adjusted.

- "AUTOCONF_H += -include $(LINUXSRC)/include/linux/kconfig.h" line is
  added avoid the following kernel module build errors indicating that
  IS_ENABLED is unintendedly undefined:

  include/asm-generic/bitops/instrumented-non-atomic.h: In function \
      '__instrument_read_write_bitop':
  include/asm-generic/bitops/instrumented-non-atomic.h:63:6: error: \
      implicit declaration of function 'IS_ENABLED' \
      [-Werror=implicit-function-declaration]
    if (IS_ENABLED(CONFIG_KCSAN_ASSUME_PLAIN_WRITES_ATOMIC)) {
        ^~~~~~~~~~

- The line that reads
    "KBUILD_EXTRA_SYMBOLS := $(BNXT_PEER_MEM_INC)/Module.symvers"
  is guarded to avoid the following build warning, which is encountered
  as BNXT_PEER_MEM_INC is not defined by default:
    WARNING: Symbol version dump "/Module.symvers" is missing.

Signed-off-by: M. Vefa Bicakci <vefa.bicakci@windriver.com>
---
 bnxt_re-220.0.12.0/Makefile | 22 +++++++++++++++-------
 1 file changed, 15 insertions(+), 7 deletions(-)

diff --git a/bnxt_re-220.0.12.0/Makefile b/bnxt_re-220.0.12.0/Makefile
index 532fb8025465..702d4a81c6f2 100644
--- a/bnxt_re-220.0.12.0/Makefile
+++ b/bnxt_re-220.0.12.0/Makefile
@@ -59,6 +59,8 @@ endif
 ifneq ($(KDIR),)
   LINUX=$(KDIR)
   LINUXSRC=$(LINUX)
+else ifneq ($(LINUX),)
+  KDIR=$(LINUX)
 endif
 
 ifeq ($(shell ls $(LINUXSRC)/include/uapi > /dev/null 2>&1 && echo uapi),)
@@ -90,26 +92,32 @@ ifeq ($(OFED_VERSION), )
 endif
 
 #find OFED version and compat-includes
-ofed_major=$(filter OFED-3.% OFED-4.%, $(OFED_VERSION))
+ofed_major=$(filter OFED-3.% OFED-4.% 5.%, $(OFED_VERSION))
 ifneq ($(ofed_major), )
-exists=$(shell if [ -e /usr/src/compat-rdma$(OFED_VERSION) ];\
+exists=$(shell if [ -e /usr/src/ofa_kernel-$(OFED_VERSION) ];\
                 then echo y; fi)
 ifeq ($(exists), )
-$(shell ln -s /usr/src/compat-rdma\
-         /usr/src/compat-rdma$(OFED_VERSION))
+$(shell ln -s /usr/src/ofa_kernel\
+         /usr/src/ofa_kernel-$(OFED_VERSION))
 endif
-OFA_BUILD_PATH=/usr/src/compat-rdma$(OFED_VERSION)
-OFA_KERNEL_PATH=/usr/src/compat-rdma$(OFED_VERSION)
+OFA_BUILD_PATH=/usr/src/ofa_kernel/default
+OFA_KERNEL_PATH=/usr/src/ofa_kernel/default
 EXTRA_CFLAGS += -DOFED_3_x
-ofed_4_17_x=$(filter OFED-4.17%, $(ofed_major))
+ofed_4_17_x=$(filter OFED-4.17% 5.%, $(ofed_major))
 ifneq ($(ofed_4_17_x), )
 EXTRA_CFLAGS += -D__OFED_BUILD__
 endif
 EXTRA_CFLAGS += -include $(OFA_KERNEL_PATH)/include/linux/compat-2.6.h
 
 AUTOCONF_H = -include $(shell /bin/ls -1 $(LINUX)/include/*/autoconf.h 2> /dev/null | head -1)
+AUTOCONF_H += -include $(LINUXSRC)/include/linux/kconfig.h
 endif #end non 3.x OFED
+
+ifneq ($(BNXT_PEER_MEM_INC),)
 KBUILD_EXTRA_SYMBOLS := $(BNXT_PEER_MEM_INC)/Module.symvers
+else
+KBUILD_EXTRA_SYMBOLS :=
+endif
 
 ifeq (OFED-NATIVE, $(findstring OFED-NATIVE, $(OFED_VERSION)))
 ifeq ($(KERNELRELEASE),)
-- 
2.25.1

